File: 백준/Silver/1991. 트리 순회/트리 순회.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *


input = stdin.readline

n = int(input())

tree = defaultdict(tuple)

for _ in range(n):
    
    parent,left,right = input().split()
    
    tree[parent] = (left,right)
    

resultPre = []
resultIn = []
resultPost = []

def preOrder(root:str):
    
    global resultPre
    
    if root == ".":
        return 

    child = tree[root]
    left = child[0]
    right = child[1]
    
    resultPre.append(root)
    preOrder(left)
    preOrder(right)

def inOrder(root:str):
    
    global resultIn
    
    if root == ".":
        return 

    child = tree[root]
    left = child[0]
    right = child[1]
    
    inOrder(left)
    resultIn.append(root)
    inOrder(right)

def postOrder(root:str):
    
    global resultPost
    
    if root == ".":
        return 

    child = tree[root]
    left = child[0]
    right = child[1]
    
    postOrder(left)
    postOrder(right)
    resultPost.append(root)

preOrder("A")
inOrder("A")
postOrder("A")

print("".join(resultPre))
print("".join(resultIn))
print("".join(resultPost))
        
    

File: 백준/Silver/1991. 트리 순회/트리 순회.swift
import Foundation


let n = Int(readLine()!)!

var adj:[String:[String]] = .init()

for _ in 0..<n {
    let nodes = readLine()!.split{$0 == " "}.compactMap{ String($0) }
    
    adj[nodes[0]] = [nodes[1],nodes[2]]
    
}

var result:String = ""

func preOrder(_ node:String) {
    
    if node == "." {
        return
    }
    result.append(node)
    preOrder(adj[node]?[0] ?? ".") // 왼쪽
    preOrder(adj[node]?[1] ?? ".") // 오른쪽
    
}

func inOrder(_ node:String) {
    
    if node == "." {
        return
    }
    inOrder(adj[node]?[0] ?? ".") // 왼쪽
    result.append(node)
    inOrder(adj[node]?[1] ?? ".") // 오른쪽
}

func postOrder(_ node:String) {
    
    if node == "." {
        return
    }
    postOrder(adj[node]?[0] ?? ".") // 왼쪽
    postOrder(adj[node]?[1] ?? ".") // 오른쪽
    result.append(node)
}



preOrder("A")
print(result)
result = ""
inOrder("A")
print(result)
result = ""
postOrder("A")
print(result)

